### ISSUE TITLE
Using reserved name for parameter name causes a parser failure

### ISSUE THREAD

**Issue by wallymathieu** [cake-build/cake#1260]
I've made the following code available:

https://github.com/wallymathieu/SemVer.FromAssembly

It's loosely based on code in elm-package and f# core. It could be used by referencing the nuget (or by porting it to c#).

Thoughts?

**Comment by gep13** [comment]
@wallymathieu are you meaning that you would like to create some aliases that could be used in Cake for this tool? If so, I think it would make a good candidate for an addin.

**Comment by wallymathieu** [comment]
Making it an addin sounds like a good plan. What kind of things would make sense to do in order for it to be a good addin?

**Comment by gep13** [comment]
@wallymathieu ideally, you would want to expose as much of the API surface for the exe as would make sense for an automated build.  You can see the source for a number of existing addin's here:

https://github.com/cake-contrib

I would suggest having a look at though, and then let us know if you have any questions.

**Comment by wallymathieu** [comment]
I've done a first take on it based on Cake.Gem:

https://github.com/wallymathieu/Cake.SemVer.FromAssembly

**Comment by wallymathieu** [comment]
I get the following error while trying out the tool:

Error: (1960,113): error CS1001: Identifier expected

I've tried to match the structure of the Cake.Gem addin.

**Comment by gep13** [comment]
@wallymathieu can you provide a list of steps on how you get to that point?  Perhaps with that, we can investigate.

**Comment by wallymathieu** [comment]
I've added Cake.SemVer.FromAssembly (version 0.0.2) in packages.config in tools. Then :

```
#addin Cake.SemVer.FromAssembly
```

Probably I'm doing something wrong, I'm not that fluent with cake yet ;)

**Comment by devlead** [comment]
@wallymathieu  Issue confirmed seems to be something with code generation and with the word Assembly.
This fails
`#addin nuget:?package=Cake.SemVer.FromAssembly`
whereas this doesn't generate a compiler error
`#addin nuget:?package=Cake.SemVer.FromAssembly_`
but obviously doesn't find package.

So this will need some investigating.

**Comment by wallymathieu** [comment]
Should we rename the package in order to avoid the issue? Cake.SemVer.UsingDiff or something else?

**Comment by gep13** [comment]
@wallymathieu just wanted to touch base on this issue...

Were you able to move forward with it?

**Comment by wallymathieu** [comment]
No, the assembly name was kind of a showstopper. I could publish the tool with another name, but I'm not sure what to call it. Any suggestions @gep13 ?

**Comment by gep13** [comment]
@wallymathieu ultimately, this is an issue in Cake, but not sure when this will be resolved, so in the mean time, a change of name would seem to be the obvious course of action.  In terms of name, what about `Cake.SemVer.FromBinary`?

**Comment by wallymathieu** [comment]
Sounds good! Should we close this issue and create a new one for the "Assembly" bug?

**Comment by wallymathieu** [comment]
Or rename the title?

**Comment by gep13** [comment]
>@wallymathieu said...
Or rename the title?

Yip, renaming the issue will be fine.

**Comment by wallymathieu** [comment]
When running the following script [test.cake](https://github.com/wallymathieu/Cake.SemVer.FromAssembly/blob/integration/test.cake) 
I get 
```
>.\tools\Cake\Cake.exe .\test.cake -target=default -verbosity=verbose
Analyzing build script...
Processing build script...
Installing addins...
Compiling build script...
Error: (1960,111): error CS1001: Identifier expected
```

**Comment by kcamp** [comment]
@wallymathieu try bumping the verbosity up to `-Verbosity Diagnostic`
That will give you a more specific error.  It appears to be an issue with the actual compilation of the script aliases.

My environment reproduces similar to yours, albeit with different line numbers.

```
Error: Roslyn.Compilers.CompilationErrorException: (2080,111): error CS1001: Identifier expected
   at Roslyn.Scripting.CommonScriptEngine.CompilationError(DiagnosticBag localDiagnostics, DiagnosticBag diagnostics)
   at Roslyn.Scripting.CommonScriptEngine.Compile(String code, String path, DiagnosticBag diagnostics, Session session, Type delegateType, Type returnType, CancellationT
oken cancellationToken, Boolean isInteractive, Boolean isExecute, CommonCompilation& compilation, Delegate& factory)
   at Roslyn.Scripting.CommonScriptEngine.Execute[T](String code, String path, DiagnosticBag diagnostics, Session session, Boolean isInteractive)
   at Roslyn.Scripting.Session.Execute(String code)
   at Cake.Scripting.Roslyn.Stable.DefaultRoslynScriptSession.Execute(Script script) in C:\git\cake\src\Cake\Scripting\Roslyn\Stable\DefaultRoslynScriptSession.cs:line 3
3
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath, IDictionary`2 arguments)
   at Cake.Commands.BuildCommand.Execute(CakeOptions options) in C:\git\cake\src\Cake\Commands\BuildCommand.cs:line 34
   at Cake.CakeApplication.Run(CakeOptions options) in C:\git\cake\src\Cake\CakeApplication.cs:line 46
   at Cake.Program.Main() in C:\git\cake\src\Cake\Program.cs:line 75
```
By stepping into `Cake.Scripting.Roslyn.RoslynCodeGenerator` you can see the code .. 
```
line 2080
public Cake.SemVer.FromBinary.Magnitude SemVerMagnitude(Cake.Core.IO.FilePath original, Cake.Core.IO.FilePath new)
{
    return Cake.SemVer.FromBinary.SemVerAliases.SemVerMagnitude(Context, original, new);
}


public Cake.SemVer.FromBinary.Magnitude SemVerMagnitude(Cake.Core.IO.FilePath original, Cake.Core.IO.FilePath new, Cake.Core.IO.FilePath output)
{
    return Cake.SemVer.FromBinary.SemVerAliases.SemVerMagnitude(Context, original, new, output);
}
```

**Comment by devlead** [comment]
`new` is a reserved keyword so probably a bad parameter name, I.e. something like `newPath` would sort it.

**Comment by kcamp** [comment]
@wallymathieu the parameter name is almost certainly the problem.  I recall reflecting on your assembly in looking at this earlier and seeing that you were using `@new` as the [escaped] parameter name.  The script alias generation does not do any reserved word filtering to escape parameter names like that.  Change that parameter name and you should be good.   Additionally, I have to wonder if that was the underlying cause of the initial issue..

From @devlead on 10/21
```
This fails
#addin nuget:?package=Cake.SemVer.FromAssembly
whereas this doesn't generate a compiler error
#addin nuget:?package=Cake.SemVer.FromAssembly_
but obviously doesn't find package.
```

**_but obviously doesn't find package._** --- this part is key to why it worked .. since it wasn't able to find the package, it doesn't load it and never attempts to generate the script method aliases..

**Comment by kcamp** [comment]
@devlead is it worth creating an issue for improvement with regard to the script method generation?  It would be easy enough to create a list of reserved keywords to filter against and properly escape the name if we wanted for the library to be fully CLR compliant.

**Comment by wallymathieu** [comment]
I've renamed this issue to match what you found. I've also updated the library.

**Comment by wallymathieu** [comment]
@kcamp even if it's easy enough, but will it be done? ;)

**Comment by kcamp** [comment]
I will have a PR for it at some point in the next day or so.

**Comment by devlead** [comment]
fixed by #1409
